packages/vm/src/imports.rs

```diff
@@ -95,156 +82,150 @@ mod errors {
     /// db_scan errors (-2_000_0xx)
     #[cfg(feature = "iterator")]
     pub mod scan {
-        /// An unknown error in the db_scan implementation
-        pub static UNKNOWN: i32 = -2_000_001;
         /// Invalid Order enum value passed into scan
         pub static INVALID_ORDER: i32 = -2_000_002;
     }
 
-    /// db_next errors (-2_000_1xx)
-    #[cfg(feature = "iterator")]
-    pub mod next {
-        /// An unknown error in the db_next implementation
-        pub static UNKNOWN: i32 = -2_000_101;
-        /// Iterator with the given ID is not registered
-        pub static ITERATOR_DOES_NOT_EXIST: i32 = -2_000_102;
-    }
+    // /// db_next errors (-2_000_1xx)
+    // #[cfg(feature = "iterator")]
+    // pub mod next {
+    // }
 }
 
-/// Reads a storage entry from the VM's storage into Wasm memory
-pub fn do_read<S: Storage, Q: Querier>(ctx: &mut Ctx, key_ptr: u32, value_ptr: u32) -> i32 {
-    let key = match read_region(ctx, key_ptr, MAX_LENGTH_DB_KEY) {
-        Ok(data) => data,
-        Err(VmError::RegionLengthTooBigErr { .. }) => return errors::REGION_READ_LENGTH_TOO_BIG,
-        Err(_) => return errors::REGION_READ_UNKNOWN,
+/// This macro wraps the read_region function for the purposes of the functions below which need to report errors
+/// in its operation to the caller in the WASM runtime.
+/// On success, the read data is returned from the expression.
+/// On failure, an error number is wrapped in `Ok` and returned from the function.
+macro_rules! read_region {
+    ($ctx: expr, $ptr: expr, $length: expr) => {
+        match read_region($ctx, $ptr, $length) {
+            Ok(data) => data,
+            Err(err) => {
+                return Ok(match err {
+                    VmError::RegionLengthTooBigErr { .. } => errors::REGION_READ_LENGTH_TOO_BIG,
+                    _ => errors::REGION_READ_UNKNOWN,
+                })
+            }
+        }
     };
-    let value: Option<Vec<u8>> = match with_storage_from_context::<S, Q, _, _>(ctx, |store| {
-        store
-            .get(&key)
-            .or_else(|_| make_runtime_err("Error reading from backend"))
-    }) {
-        Ok(v) => v,
-        Err(VmError::UninitializedContextData { .. }) => return errors::NO_CONTEXT_DATA,
-        Err(_) => return errors::DB_UNKNOWN,
+}
+
+#[cfg(feature = "iterator")]
+/// This macro wraps the maybe_read_region function for the purposes of the functions below which need to report errors
+/// in its operation to the caller in the WASM runtime.
+/// On success, the optionally read data is returned from the expression.
+/// On failure, an error number is wrapped in `Ok` and returned from the function.
+macro_rules! maybe_read_region {
```

Comment:
Having this marco, can we remove the function `maybe_read_region` entirely? I mean the whole logic of `maybe_read_region` does not really belong in `mod memory`. It is the caller's problem to either read memory or don't if the caller specific case nullptr was detected.

Reply:
I don't have a strong opinion either way. an argument can be made for `maybe_read_region`'s interface, which can be very useful in the future, but today it really is only used from here.

Reply:
Yeah, feel free to keep it as is. We can refine this later